incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaPublishFactory.java

```diff
@@ -108,12 +110,10 @@ public class MqttKafkaPublishFactory implements BindingHandler
     private final int kafkaTypeId;
     private final Signaler signaler;
     private final LongFunction<MqttKafkaBindingConfig> supplyBinding;
-    private KafkaKeyFW key;
-
-    private OctetsFW[] topicNameHeaders;
-    private OctetsFW clientIdOctets;
-    private String16FW binaryFormat;
-    private String16FW textFormat;
+    private final String16FW binaryFormat;
+    private final String16FW textFormat;
+    private final String kafkaTopic;
+    private final String kafkaRetainedTopic;
```

Comment:
These topic names can be stored as `StringFW` to `utf-8` encode once, rather than storing as `String` and `utf-8` encoding per usage.